### PR TITLE
Feature #2699 consensus_config

### LINKED ISSUE #2699: Refine TC-Pairs consensus diagnostics configuration options
## Describe the Enhancement ##
Issue #2476 enhances the logic in TC-Pairs to include diagnostics when computing consensus tracks. This issue is to refine that logic based on feedback from @musgrave-kate to support the 2 new configuration options described in this [issue comment](https://github.com/dtcenter/MET/issues/2476#issuecomment-1728439945).

1. Print a warning if diagnostics are present for AT LEAST one member but not present for ALL members.
2. Add two new config options in the `consensus` dictionary entries to provide finer handling of diagnostics:
  - `diag_min_req` is similar to the `min_req` entry to specify the required number of diagnostics present to be included in the output.
  - `diag_required` is similar to the `required` entry to specify the models that must contain diagnostics for them to be calculated.

In addition, revisit the diagnostics consensus test added to `unit_tc_diag.xml` using contrived data. Recommend replacing it with a better example.

### Time Estimate ###
2 days

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
None needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required: @jvigh to review.

### Labels ###
- [x] Review default **alert** labels
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Milestone and Projects ###
- [x] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [x] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- dtcenter/METplus#2430 to update the TC-Pairs wrapper.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [x] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [x] Iterate until the reviewer(s) accept and merge your changes.
- [x] Delete your fork or branch.
- [x] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>
Adds new `diag_required` and `diag_min_req` config options to the `consensus` dictionary.

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
Ran this [testing GHA](https://github.com/dtcenter/MET/actions/runs/6867002500) workflow to confirm that the code all compiles and runs as expected. It does produce a different output file, as expected, based on changes to in `unit_tc_pairs.xml`.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
  - Review the code changes on the [Files changed](https://github.com/dtcenter/MET/pull/2736/files) tab.
  - Review the documentation changes on the [Files changed](https://github.com/dtcenter/MET/pull/2736/files) tab or via the [PR-2736 Read-The-Docs Build](https://met--2736.org.readthedocs.build/en/2736/Users_Guide/tc-pairs.html).
  - Review the improved, more realistic diagnostics consensus test.

This feature branch is compiled and available for your use on seneca in:
`/d1/projects/MET/MET_pull_requests/met-12.0.0/beta2/MET-feature_2699_consensus_config`
 
- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [x] Do these changes include sufficient testing updates? **[Yes]**
Updated `unit_tc_paris.xml` test to use real data to create a consensus using GFSO, HMON, and HWRF.

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
Modifies the contents of a single TC-Pairs output file `al092022_20220926_DIAGNOSTICS_CONSENSUS.tcst` due to modified inputs.

- [x] Please complete this pull request review by **[Wed 11/15/23]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by jvigh** [comment]
@JohnHalleyGotway  Are these errors in the diff testing important:
https://github.com/dtcenter/MET/actions/runs/6894967897/job/18759196089?pr=2736

File: docs/Users_Guide/tc-pairs.rst [hunk 1]
```diff
@@ -7,7 +7,7 @@ TC-Pairs Tool
...
+The TC-Pairs tool provides verification for tropical cyclone forecasts in ATCF file format. It matches an ATCF format tropical cyclone (TC) forecast with a second ATCF format reference TC dataset (most commonly the Best Track analysis). The TC-Pairs tool processes both track and intensity adeck data and probabilistic edeck data. The adeck matched pairs contain position errors, as well as wind, sea level pressure, and distance to land values for each TC dataset. The edeck matched pairs contain probabilistic forecast values and the verifying observation values. The pair generation can be subset based on user-defined filtering criteria. Practical aspects of the TC-Pairs tool are described in :numref:`TC-Pairs_Practical-information`.
```
Comment (jvigh) [inline docs/Users_Guide/tc-pairs.rst]: What is meant by "wind" in this sentence? If it is actually storm intensity, I suggest changing "wind" to "intensity". But if it's something different (like wind radii), then it'd be good to clarify that as well. 

Similarly, is "sea level pressure" actually "minimum sea level pressure" (the lowest sea level pressure near the storm center)? If so, I'd suggest changing this as well.

**Review by jvigh — APPROVED** [review]
I have reviewed the commit history, code changes, GHA, and the full documentation chapter. Everything looks good. I've left one small note of clarification. 

I note that I did not have time to run the tools on Seneca and test edge cases. 

Nice work on this.

File: docs/Users_Guide/tc-pairs.rst (hunk 1 above)
Comment (JohnHalleyGotway) [reply docs/Users_Guide/tc-pairs.rst]: Thanks for the clarification @jvigh.

Changing the wording...
FROM: `The adeck matched pairs contain position errors, as well as wind, sea level pressure, and distance to land values for each TC dataset.`
TO: `The adeck matched pairs contain position errors, as well as values for each TC dataset, including wind speed, wind radii, minimum sea level pressure, and distance to land.`

**Comment by JohnHalleyGotway** [comment]
> @JohnHalleyGotway Are these errors in the diff testing important: [dtcenter/MET/actions/runs/6894967897/job/18759196089?pr=2736](https://github.com/dtcenter/MET/actions/runs/6894967897/job/18759196089?pr=2736)

@jvigh yes, those ERRORS are important, but they won't be addressed via this PR. They illustrate a short-coming in the existing PERL and R-based differencing logic that MET currently uses.

@natalieb-noaa will be working on an overhaul to MET's unit testing framework and differencing logic... switching over to using the METplus Python-based one, and likely enhancing it along the way. So this is definitely an example of a difference in the output that should be handled in a more elegant way!
